### ISSUE TITLE
Sneaky Tile Movement

### ISSUE THREAD

**Issue by calclavia** [NOVA-Team/NOVA-Core#143]
Should we support a seamless way to move TE equivalent things in Minecraft? Or is that too specific and should have a modder write it in a wrapper?

In MFFS/EDX, I have things that can move TileEntities. Just wondering if it's appropriate to add that support here. Opinions are welcome.

**Comment by Caellian** [comment]
And please do not tell me TEs are still considered Entities by NOVA.

**Comment by Caellian** [comment]
Is your code Minecraft dependant?
If it is, can it be made NOVA dependant?
If yes, it would be useful to add it to NOVA, otherwise you could add it to
wrapper.

**Comment by RX14** [comment]
Well, it would be abstracted as just moving blocks, whether they are TEs or not. Maybe a isMovable on the block?

**Comment by Caellian** [comment]
Actually yes, TEs are bound to blocks, so moving a block should move a
corresponding TE without need for special handling.

But keep in mind that Minecraft is probably the only voxel game storing TEs
and their positions. This probably requires a custom management system for
voxels with extended data that can later on be casted to TEs for Minecrafts
case.

**Comment by RX14** [comment]
Well, it would be more like a simple helper function that moved blocks, whether they are TEs or dumb decoration blocks.

**Comment by anti344** [comment]
There are Stateful and stateless blocks in NOVA, there are no tileentities and such, so yeah some kind of util for moving NOVA blocks (and whatever implementation of it in wrappers) should be nice

**Comment by asiekierka** [comment]
I think blocks need a way to know when they're being moved. Many of them will store absolute coordinates, after all. (The problem there is: what if you're moving a cuboid that contains a multiblock structure? What if you're moving part of a multiblock structure? How does one differentiate?)

**Comment by calclavia** [comment]
@asiekierka I guess we'll throw in events. Blocks technically should store relative values to be safe, but there's no way to enforce this (except yelling at developers through the docs :P)

Basically the API for this is simple:
```
block.setPosition(...)
```

**Comment by Caellian** [comment]
What about a boolean method in BlockComponent that is called when a block tries to move and returns true if the block can move.

For example, bedrock would always return false, if you push a dirt block, is will recursively check if block next to it can be moved (there should be a way of preventing infinite recursion).

so the method would look like this:
```java
public boolean moveBlock(direction, numOfBlockSpacesMoved, recursionCounter){
    return world.getBlock(this.pos.increment(direction, 1)).moveBlock(direction, numOfBlockSpacesMoved, ++recursionCounter)
}
```
Ignore incorrect names, this is just an example for a default method. 

@calclavia , move block and set block position are two very different things.

Setblock could also be done with a method.

This way isn't ideal, but it allows mod developers to easily change block behaviour (TP block randomly if pushed by piston).
I'm not sure if events should be used for this as it might end up being really laggy.

**Comment by ExE-Boss** [comment]
Do you mean what I implemented in #290?
